### ISSUE TITLE
Serializing and Deserializing after loading Keras model is broken

### ISSUE THREAD

**Issue by bogdanteleaga** [deeplearning4j/deeplearning4j#2474]
So I'm trying to package a model inside a .jar. The Keras model doesn't provide a filestream API and java doesn't let me use files inside a .jar as a regular file. So I decided to load the model in DL4J and reserialize it. The problem is what comes out the other end is different. Code below:

```java
    //Store the state of this neural network
    public void storeNetwork() {
        FileOutputStream fo = null;
         try {
             fo = new FileOutputStream("src/memory/speed_model.zip");
             ModelSerializer.writeModel(this.speed_model, fo, false);
             fo.close();
             fo = new FileOutputStream("src/memory/steer_model.zip");
             ModelSerializer.writeModel(this.steer_model, fo, false);
             fo.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void loadNetwork() {

        // Read from disk using FileInputStream
        FileInputStream fi = null;
        try {
            fi = new FileInputStream("src/memory/speed_model.zip");
            this.speed_model = ModelSerializer.restoreMultiLayerNetwork(fi);
            fi.close();
            fi = new FileInputStream("src/memory/steer_model.zip");
            this.steer_model = ModelSerializer.restoreMultiLayerNetwork(fi);
            fi.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void loadNets() {
        this.speed_model = this.loadNet("speed");
        this.steer_model = this.loadNet("steering");
    }

    // Load a neural network from memory
    private MultiLayerNetwork loadNet(String modelName) {
        MultiLayerNetwork model = null;

        try {
            String tmpDir = System.getProperty("java.io.tmpdir");
            String modelHDF5 = modelName + "_nn.hdf5";
            String modelFilePath = tmpDir + modelHDF5;
            File modelFile = new File(modelFilePath);
            FileUtils.copyInputStreamToFile(
                    getClass().getResourceAsStream(modelHDF5),
                    modelFile);
            model = org.deeplearning4j.nn.modelimport.keras.Model.importSequentialModel(
                    modelFilePath);
            modelFile.delete();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return model;
    }
```

You can also see one workaround for dealing with the keras API issue, but that doesn't work everywhere.

So I'm using `loadNets` then `storeGenome` the first time around. Once I have the models, I'm using `loadGenome` only to load them. The resulting model is not completely different, meaning it performs "well" but it's definitely different. I'm using them for TORCS right now and while the initial model can finish a circuit the loaded one gets stuck quite early on, but it's not completely hopeless. I imagine just *some* of the information gets lost, but I can't exactly put my finger on what.

**Comment by raver119** [comment]
To use model bundled into jar, we have ClassPathResource available in DataVec. It does exactly what you want transparently extracting your file from resources.

**Comment by raver119** [comment]
Also need more details, how "different" are saved/restored models.

I.e. what says comparison for saved/restored model params?

**Comment by agibsonccc** [comment]
@turambar this is relevant for you I believe.

**Comment by turambar** [comment]
@agibsonccc Got it. @bogdanteleaga Is the originally imported Keras model (_before_ being serialized and re-serialized) performing properly?

**Comment by turambar** [comment]
@bogdanteleaga Got any further details for us? Otherwise, we're not going to be able to help!

**Comment by bogdanteleaga** [comment]
Nope, had a pretty stressing deadline for this and since the network wasn't *that* complicated and I did not especially need dl4j I ended up coding it myself in pure Java.

But before that a couple more things happened: I realized that if I load from Keras -> save to disk -> reload from disk in the same run of the program then it works. However it didn't work when I just read from the disk. Moreover, running the .jar instead of running the app inside intellij produced different results not only on my machine but on others as well. Unfortunately, my understanding of Java and especially the build system is quite atrocious and I believe that might have been the root cause here, even though I cannot explain all of the little quirks I've noticed just based on this.

I'm sorry, but I cannot seem to be able to reproduce this in a sane manner at this point in time and it's not a priority anymore either, so I'll close the issue. Ping me on gitter these days while it's fresh in my head if you'd like more information.
